File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for c in s:
            match c:
                case '(' | '[' | '{':
                    stack.append(c)
                case ')':
                    if not stack or stack.pop(-1) != '(':
                        return False
                case ']':
                    if not stack or stack.pop(-1) != '[':
                        return False
                case '}':
                    if not stack or stack.pop(-1) != '{':
                        return False
        return not stack

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        for(int i = 0; i < s.length(); i++) {
            if(s[i] == '(' or s[i] == '{' or s[i] == '[') {
                st.push(s[i]);
            }
            else if(st.empty()) {
                return false;
            }
            else {
                if(s[i] == ')') {
                    if(st.top() == '(') {
                        st.pop();
                    }
                    else return false;
                }
                else if(s[i] == ']') {
                    if(st.top() == '[') {
                        st.pop();
                    }
                    else return false;
                }
                else {//if(s[i] == '}') {
                    if(st.top() == '{') {
                        st.pop();
                    }
                    else return false;
                }
            }
        }
        return st.empty();
    }
};